src/services/textChanges.ts

```diff
@@ -590,104 +586,82 @@ namespace ts.textChanges {
             });
         }
 
-        public getChanges(): FileTextChanges[] {
+        /**
+         * Note: after calling this, the TextChanges object must be discarded!
+         * @param validate only for tests
+         *    The reason we must validate as part of this method is that `getNonFormattedText` changes the node's positions,
+         *    so we can only call this once and can't get the non-formatted text separately.
+         */
+        public getChanges(validate?: ValidateNonFormattedText): FileTextChanges[] {
             this.finishInsertNodeAtClassStart();
-            return group(this.changes, c => c.sourceFile.path).map(changesInFile => {
+            return changesToText.changesToText(this.changes, this.newLineCharacter, this.formatContext, validate);
+        }
+    }
+
+    export type ValidateNonFormattedText = (node: Node, text: string) => void;
+
+    namespace changesToText {
+        export function changesToText(changes: ReadonlyArray<Change>, newLineCharacter: string, formatContext: formatting.FormatContext, validate: ValidateNonFormattedText): FileTextChanges[] {

```

Comment:
I find this function name unhelpful.  I'm guessing "getChangesToText" is closer to its intended behavior.

Reply:
How about `textChangesFromChanges`?

Reply:
I'm just surprised there's no verb.